btcec/signature.go

```diff
@@ -396,3 +401,134 @@ func RecoverCompact(curve *KoblitzCurve, signature,
 
 	return key, ((signature[0] - 27) & 4) == 4, nil
 }
+
+var one = big.NewInt(1)
+
+// SignRFC6979 generates a deterministic ECDSA signature according to RFC 6979 and BIP 62.
+func SignRFC6979(privateKey *PrivateKey, hash []byte) (*Signature, error) {
+
+	privkey := privateKey.ToECDSA()
+	N := order
+	k := NonceRFC6979(privkey.D.Bytes(), hash)
+	inv := new(big.Int).ModInverse(k, N)
+	r, _ := privkey.Curve.ScalarBaseMult(k.Bytes())
+	r.Mod(r, N)
+
+	if r.Sign() == 0 {
+		return nil, errors.New("calculated R is zero")
+	}
+
+	e := hashToInt(hash, privkey.Curve)
+	s := new(big.Int).Mul(privkey.D, r)
+	s.Add(s, e)
+	s.Mul(s, inv)
+	s.Mod(s, N)
+
+	if s.Cmp(halforder) == 1 {
+		s = new(big.Int).Sub(order, s)
+	}
+
+	return &Signature{R: r, S: s}, nil
+}
+
+// NonceRFC6979 generates a ECDSA nonce (`k`) deterministically according to RFC 6979.
+// It takes a 32-byte hash as an input and returns 32-byte nonce to be used in ECDSA algorithm.
+func NonceRFC6979(privkey []byte, hash []byte) *big.Int {
+
+	q := order
+	x := new(big.Int).SetBytes(privkey)
+	alg := fastsha256.New
+
+	qlen := q.BitLen()
+	holen := alg().Size()
+	rolen := (qlen + 7) >> 3
+	bx := append(int2octets(x, rolen), bits2octets(hash, q, qlen, rolen)...)
+
+	// Step B
+	v := bytes.Repeat([]byte{0x01}, holen)
+
+	// Step C
+	k := bytes.Repeat([]byte{0x00}, holen)
```

Comment:
This can be optimized since Go promises to automatically zeroes all allocated memory using the much faster `runtime·memclr`.

```Go
// Step C.  Go promises to automatically zero all allocated memory.
k := make([]byte, holen)
```